src/librustc_codegen_ssa/mir/constant.rs

```diff
@@ -46,7 +50,10 @@ impl<'a, 'tcx, Bx: BuilderMethods<'a, 'tcx>> FunctionCx<'a, 'tcx, Bx> {
                     instance,
                     promoted: None,
                 };
-                self.cx.tcx().const_eval(ty::ParamEnv::reveal_all().and(cid))
+                self.cx.tcx().const_eval(ty::ParamEnv::reveal_all().and(cid)).map_err(|err| {
+                    self.cx.tcx().sess.span_err(constant.span, "erroneous constant encountered");
```

Comment:
This is an additional fix for when users turn off all lints with `allow`. In that case they may get a compiling program that runs into an abort. We now report a hard error again, as we already do in the monomorphized case

Reply:
Wait, I thought we *deliberately* don't make `const_err` a hard error? Because that would be trivial to do, but there are backcompat concerns or so. So why don't these concerns apply here?

Reply:
Also, won't this lead to errors being emitted in the middle of a Miri (the tool) execution? That seems odd.^^

Reply:
This is `codegen_ssa`, and we already have a hard error for using erroneous constants in runtime code, it just doesn't work for polymorphic code because it's implemented on MIR. All this PR does is make the hard error work for polymorphic code, too.

Reply:
See https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=afc895c5754a7963154193ad32702bb1

If you comment out the `let x = FOO;` you don't get a hard error.

Reply:
> This is codegen_ssa

Good point, somehow this looked so much like Miri.^^

But why does codegen_ssa ever run on polymorphic code...?

Reply:
it doesn't. It actually monorophizes it. But the MIR is polymorphic, so we can't detect the problem in MIR. That's the whole point of this PR: fail at monomorphization time.

Reply:
But for monomorphization we already errored before, as you example shows.

Looks to me like we are monomorphizing constants in (at least) two places, and they used to be inconsistent wrt. error reporting? And this PR makes them less inconsistent? Any chance those could use the same code to avoid inconsistencies entirely?

Reply:
The difference is that nonpolymorphic functions do not get monomorphized (because they don't need to get monomorphized). So when we const prop a nonpolymorphic function, when we encounter a constant, we can just eval it. When we encounter a constant in a polymorhphic function, we may only be able to eval it at monomorphization time.

Reply:
If we monomorphized all MIR before codegenning it, this would solve itself, but we'd require loads of CPU and memory to do that, so we do the monomorphization during codegen on the fly.

Reply:
Wait so the hard error from [your example](https://github.com/rust-lang/rust/pull/67134#discussion_r355413288) does *not* come from codegen, but from const-prop? That is surprising.

Reply:
Yes, because const prop can actually eliminate the constant entirely, so we need to report it there. Same problem as the DCE problem discussed in the other thread. We can unify the scheme in one go.

Reply:
Ah, so both of my questions have the same answer? I love it when that happens. :D